src/libraries/System.Text.Json/System.Text.Json.SourceGeneration.UnitTests/TypeWrapperTests.cs

```diff
@@ -0,0 +1,148 @@
+                public static partial class ExternType { }
+              }";
+
+            MetadataReference[] additionalReferences = { MetadataReference.CreateFromImage(referencedImage) };
+
+            // Compilation using the referenced image should fail if out MetadataLoadContext does not handle.
+            Compilation compilation = CompilationHelper.CreateCompilation(source, additionalReferences);
+
+            JsonSerializerSourceGenerator generator = new JsonSerializerSourceGenerator();
+
+            Compilation newCompilation = CompilationHelper.RunGenerators(compilation, out var generatorDiags, generator);
+
+            // Make sure compilation was successful.
+            Assert.Empty(generatorDiags);
+            Assert.Empty(newCompilation.GetDiagnostics());
+
+            // Should find both types since compilation above was successful.
+            Assert.Equal(2, generator.foundTypes.Count);
+        }
+
+        [Fact]
+        public void CanGetAttributes()
+        {
+            string source = @"
+            using System;
+            using System.Text.Json.Serialization;
+
+              namespace HelloWorld
+              {
+                [JsonSerializable]
+                public class MyType {
+
+                    [JsonInclude]
+                    public double PublicDouble;
+                    [JsonPropertyName(""PPublicDouble"")]
+                    public char PublicChar;
+                    [JsonIgnore]
+                    private double PrivateDouble;
+                    private char PrivateChar;
+
+                    public MyType() {{ }}
+                    [JsonConstructor]
+                    public MyType(double d) {{ PrivateDouble = d; }}
+
+                    [JsonPropertyName(""TestName"")]
+                    public int PublicPropertyInt { get; set; }
+                    [JsonExtensionData]
+                    public string PublicPropertyString { get; set; }
+                    [JsonIgnore]
+                    private int PrivatePropertyInt { get; set; }
+                    private string PrivatePropertyString { get; set; }
+
+                    [Obsolete(""Testing"", true)]
+                    public void MyMethod() { }
+                    public void MySecondMethod() { }
+                }
+              }";
+
+            Compilation compilation = CompilationHelper.CreateCompilation(source);
+
+            JsonSerializerSourceGenerator generator = new JsonSerializerSourceGenerator();
+
+            Compilation outCompilation = CompilationHelper.RunGenerators(compilation, out var generatorDiags, generator);
+
+            // Check base functionality of found types.
+            Assert.Equal(1, generator.foundTypes.Count);
+            Assert.Equal("HelloWorld.MyType", generator.foundTypes["MyType"].FullName);
+
+            Type foundType = generator.foundTypes.First().Value;
+
+            // Check for ConstructorInfoWrapper attribute usage.
```

Comment:
Will use a list of tuples List<(string nodeName, string[] attributeNames)>.
Dictionaries won't work because if method overload or constructor overloads, key repeats.